analyses/upstream-analysis/02_run_SoupX.Rmd

```diff
@@ -0,0 +1,326 @@
+
+If contamination fraction estimate is > 20%, it is recommended to set `rho` manually. In this case, we will set 5% background to be removed by default.
+
+```{r run-SoupX, echo=TRUE}
+# Create contamination summary table
+contamination_summary <- matrix(ncol = 4, nrow = 1) %>% as.data.frame()
+colnames(contamination_summary) <- c("sample_name", "rho_estimated_by_SoupX", "rho_adjusted", "rho_removed")
+
+for (i in seq_along(sample_name)) {
+  
+  cat("Beginning to process sample", sample_name[i], "\n")
+  cat("Reading in data\n")
+  
+  filtered_matrix_file <- file.path(data_dir, sample_name[i], "outs", "filtered_feature_bc_matrix.h5") 
+  raw_matrix_file <- file.path(data_dir, sample_name[i], "outs", "raw_feature_bc_matrix.h5") 
+  clusters_file <- file.path(data_dir, sample_name[i], "outs", "analysis", "clustering", "gene_expression_graphclust", "clusters.csv") 
+
+  # Read data
+  filtered_matrix <- Seurat::Read10X_h5(filtered_matrix_file, use.names = TRUE)
+  raw_matrix <- Seurat::Read10X_h5(raw_matrix_file, use.names = TRUE)
+  clusters <- read.csv(clusters_file)
+  
+  # Save plot
+  #pdf(file = paste0(SoupX_plots_dir, "/", sample_name[i], "_SoupX_contamination_fraction_plot.pdf", sep = ""))
+  
+  # Create df
+  mDat <- data.frame(clusters=clusters$Cluster, row.names=clusters$Barcode)
+  
+  # Make a “SoupChannel”, the object needed to run SoupX
+  sc_raw <- SoupChannel(raw_matrix, filtered_matrix, mDat)
+
+  # Estimate contamination fraction rho
+  sc <- autoEstCont(sc_raw, forceAccept = TRUE)
+  print(sc)
+  soup_fraction_value <- sc$fit$rhoEst
+  
+  # To not remove more than 20% background
+  if (soup_fraction_value > 0.2) {
+    print(soup_fraction_value)
+    cat("Estimated soup fraction >20%, set to 5% by default.\n")
+    sc = setContaminationFraction(sc_raw, soup_fraction_value_default)
+    contamination_summary[i,] <- c(sample_name[i], soup_fraction_value, "yes", soup_fraction_value_default)
+    print(sc)
+    } else {
+      cat("Estimated soup fraction\n is ", `soup_fraction_value`, "\n")
+      contamination_summary[i,] <- c(sample_name[i], soup_fraction_value, "no", soup_fraction_value)}
+   
+  # Correcting expression profile and clean data
+  # This will contain a corrected matrix to be used in place of the original table of counts in downstream analyses.
+  out <- adjustCounts(sc, roundToInt = TRUE) # roundToInt option to make sure we output integer matrix
+  colnames(out) <- paste0(sample_name, ":", colnames(out))
+  colnames(out) = gsub('-1$','', colnames(out)) # remove -1
+
+  # Let's save corrected matrix as a separate assay
+  seurat_obj  <- CreateSeuratObject(counts = out)
+  seurat_obj[["RNA_SoupX"]] <- CreateAssayObject(counts = seurat_obj@assays$RNA@counts)
+  seurat_obj@assays$RNA@counts <- out
```

Comment:
Mostly for my understanding, what is the reason for creating the Seurat object with counts = out, then resetting seurat_obj@assays$RNA@counts to out?

Reply:
I see that the RNA_SoupX assay is the SoupX corrected matrix. But I don't follow how `seurat_obj@assays$RNA@counts <- out` is saving the raw matrix.

Doesn't the `out` matrix only have the SoupX corrected counts, based on how it is created by the code in line 264: `out <- adjustCounts(sc, roundToInt = TRUE)`? So `seurat_obj@assays$RNA@counts <- out` would also be saving the corrected matrix in the RNA assay?

Reply:
This code chunk has been removed now as it is taken care of in the `04_run_filter_object.Rmd` script. See fb4dace73b026de44aeaa80e8060b3611558e8ed.